src/authInfo.ts

```diff
@@ -1085,12 +1087,44 @@ export class AuthInfo extends AsyncCreatable<AuthInfo.Options> {
     try {
       this.logger.info(`Sending request for Username after successful auth code exchange to URL: ${url}`);
       const response = await new Transport().httpRequest({ url, headers });
-      return asString(parseJsonMap(response.body).Username);
+      if (response.statusCode >= 400) {
+        this.throwUserGetException(response as never);
+      } else {
+        return asString(parseJsonMap(response.body).Username);
+      }
     } catch (err) {
       throw SfdxError.create('@salesforce/core', 'core', 'AuthCodeUsernameRetrievalError', [orgId, err.message]);
     }
   }
 
+  /**
+   * Given an error while getting the User object, handle different possibilities of response.body.
+   *
+   * @param response
+   * @private
+   */
+  private throwUserGetException(response: never) {
+    let messages = '';
+    let body: AnyJson | string = getString(
+      response,
+      'body',
+      JSON.stringify({ message: 'UNKNOWN', errorCode: 'UNKNOWN' })
+    );
+    try {
+      body = parseJson(getString(response, 'body', JSON.stringify({ message: 'UNKNOWN', errorCode: 'UNKNOWN' })));
+      if (isArray(body)) {
+        messages = body
+          .map((line) => getString(line, 'message') ?? getString(line, 'errorCode', 'UNKNOWN'))
+          .join(os.EOL);
+      } else {
+        messages = getString(body, 'message') ?? getString(body, 'errorCode', 'UNKNOWN');
+      }
+    } catch (err) {
+      messages = `${body}`;
```

Comment:
If parse fails, just use string representation of body. Server errors >= 500 may return an html block instead of JSON, so the message will be ugly, but will at least provide the needed info to user.

Reply:
Ya, those html pages being returned from the server are annoying. We can have follow-ups work items to clean those up if we start getting a bunch of them.